### PR TITLE
Fix early-abort handling in array_parse

### PR DESCRIPTION
**Pull request by pfitzseb**
Fixes https://github.com/JuliaLang/JuliaSyntax.jl/issues/66.

### CODE REVIEW

File: src/parser.jl [hunk 1]
```diff
@@ -2617,6 +2617,9 @@ function parse_array(ps::ParseState, mark, closer, end_is_symbol)
     # For an excellent overview of Pratt parsing, see
     # https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html
     (dim, binding_power) = parse_array_separator(ps, array_order)
+    if binding_power == typemin(Int)
+        return (K"hcat", EMPTY_FLAGS)
```
Comment (c42f) [inline src/parser.jl:2621]: Let's consume `]` if it's present.

```suggestion
        bump_closing_token(ps, closer)
        return (K"hcat", EMPTY_FLAGS)
```

**Review by c42f — COMMENTED** [review]
Let's ensure `parse_array` looks for the closer for consistency.  Still doesn't give the best error, but this does seem like an obscure edge case.

BTW, there's a tool `itest_parse` for interactive development showing the various trees and the flisp parser output

```
julia> include("test/test_utils.jl")

julia> itest_parse(JuliaSyntax.parse_expr, "[a@b]")
# Code:
[a@b]

# Green tree:
     1:5      │[hcat]
     1:1      │  [                          "["
     2:2      │  Identifier             ✔   "a"
     3:4      │  [error]                 ✘
     3:3      │    @                    ✔   "@"
     4:4      │    Identifier           ✔   "b"
     5:5      │  ]                          "]"
Error: Expected `]`
@ line 1:3
[a@b]

# SyntaxNode:
(hcat a (error-t ✘ b))
```

**Review by c42f — APPROVED** [review]

File: src/parser.jl [hunk 2]
```diff
@@ -2617,6 +2617,10 @@ function parse_array(ps::ParseState, mark, closer, end_is_symbol)
     # For an excellent overview of Pratt parsing, see
     # https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html
     (dim, binding_power) = parse_array_separator(ps, array_order)
+    if binding_power == typemin(Int)
+        bump_closing_token(ps, closer)
```
Comment (c42f) [inline src/parser.jl:2621]: Oh let's also add the case that can lead to this branch inline. It makes reading the code later a *lot* easier

```suggestion
        # [x@y  ==>  (hcat x (error-t ✘ y))
        bump_closing_token(ps, closer)
```

Comment (c42f) [reply src/parser.jl:2621]: My plan has kind of been to extract these inline comments automatically and use them as test cases ... haven't got around to that yet though

Comment (pfitzseb) [reply src/parser.jl:2622]: Yeah, I saw that comment in the tests. Will be a bit annoying to keep the inline annotations in sync with the tests until that's done.
